src/System.Management.Automation/engine/Utils.cs

```diff
@@ -1748,9 +1743,7 @@ internal static void NotNullOrEmpty(string value, string paramName)
 
         internal static void NotNullOrEmpty(ICollection value, string paramName)
         {
-            ArgumentNullException.ThrowIfNull(value, paramName);
-
-            if (value.Count == 0)
+            if (value is null || value.Count == 0)
```

Comment:
As I pointed already in another PR, this generates worst code.
Also in .Net 8 Preview 1 we will get ArgumentOutOfRangeException.ThrowIfZero (and more) so that the code will be:
```c#
ArgumentNullException.ThrowIfNull(value, paramName);
ArgumentNullException.ThrowIfZero(value.Count);
```

Reply:
You are right it will throw a different exception, I don't know if it will be a problem (maybe we just have to change the tests).

From https://github.com/dotnet/runtime/issues/68339
`public static void ThrowIfZero(int value, string paramName);` it seems can set any string as parameter name

Reply:
The original code is not correct. There are not too few places in our code where an exception is called that is not what it should be. This is not even a breaking change, as it is not a _functional_ exception, and it is worth correcting it for the correct one. If someone even uses this behavior in their code, it is catastrophically bad code.

> I don't know what will be used as the parameter name, "value.count" I guess?

Yes, as the attribute name (CallerArgumentExpression) says `expression` is used as "name". It is more informative and more correct. (Otherwise this API would never have been approved.)

So I think we should fix such protective (non-functional) exceptions to the correct ones in whole code base.

Reply:
> So I think we should fix such protective (non-functional) exceptions to the correct ones in whole code base.

That will need to be discussed and approved by all maintainers.

`ArgumentOutOfRangeException` won't be the right exception either. If you insist to be accurate about the exception, then an `ArgumentException` with a customized error message should be the way to go.

In my personal view, changing those exceptions is not worth the time (discussion, making the change, and code review). But if all other maintainers agree to make the change, then I'm fine to replace the current code with a customized `ArgumentException`.

Reply:
I've submitted a PR for `ThrowIfNullOrEmpty(String)` with the breaking changes described (#19197).

Reply:
And what is the difference from #19197?

Reply:
Here is the difference:
> `ArgumentOutOfRangeException` won't be the right exception either. If you insist to be accurate about the exception, then an ArgumentException with a customized error message should be the way to go.

And even though I don't think it worth the effort, but
> if all other maintainers agree to make the change, then I'm fine to replace the current code with a customized `ArgumentException`.

Reply:
I looked at the examples in .Net and they use ArgumentOutOfRangeException.ThrowIfZero in similar cases. So I don't understand why it has to be an ArgumentException.

Ex. https://source.dot.net/#System.Net.WebSockets.Client/System/Net/WebSockets/ClientWebSocketOptions.cs,227

Reply:
Is it a correct usage of `ArgumentOutOfRangeException.ThrowIfZero`? I don't think so.
The error message of the exception thrown from `[System.ArgumentOutOfRangeException]::ThrowIfZero[int](0, "paramName")` is:
> 'paramName' must be a non-zero value. (Parameter 'paramName')

In this case and our case here, the passed in value is a collection, and of course it's a non-zero value.

If you don't specify an argument for the `paramName` parameter, then the parameter shown in the error is simply wrong -- there is no such parameter called `'paramName.Count'`.

Reply:
> I looked at the examples in .Net and they use ArgumentOutOfRangeException.ThrowIfZero in similar cases. So I don't understand why it has to be an ArgumentException.
> 
> Ex. https://source.dot.net/#System.Net.WebSockets.Client/System/Net/WebSockets/ClientWebSocketOptions.cs,227

The example is arguably incorrect, and this pattern is not widespread:

https://github.com/search?q=repo%3Adotnet%2Fruntime+%2FArgument.%2BException%5C.Throw.%2B%5C.Count%2F&type=code

Reply:
> this pattern is not widespread

A day after public release? 😺 

> IMO, if there is a new API that check on empty collection, then that will be a good fit here, otherwise, a custom exception should be used if the purpose is to be accurate.

I found .Net rejected such API proposal. https://github.com/dotnet/runtime/issues/62628#issuecomment-1005102641

Reply:
> > So I think we should fix such protective (non-functional) exceptions to the correct ones in whole code base.
> 
> That will need to be discussed and approved by all maintainers. If you insist on fixing those exceptions, then please open an issue on it and mark it with `Review - Maintainer`.
> 
> `ArgumentOutOfRangeException` won't be the right exception either. If you insist to be accurate about the exception, then an `ArgumentException` with a customized error message should be the way to go.
> 
> In my personal view, changing those exceptions is not worth the time (discussion, making the change, and code review). But if all other maintainers agree to make the change, then I'm fine to replace the current code with a customized `ArgumentException`.

Done #19212

Reply:
> A day after public release? 😺

It doesn't matter. The use you referenced is an incorrect usage of that API.

> Done https://github.com/PowerShell/PowerShell/issues/19212

Thanks. The next maintainer meeting will be the next Tuesday.

Reply:
> > A day after public release? 😺
> 
> It doesn't matter. The use you referenced is an incorrect usage of that API.

I took a look at the history of this API. It was created by Stephan Toub and updated by him. It is hard to suspect him of negligence. I think this is intentional since the exact API was rejected as I mentioned above.
So after a while this pattern may become a frequently used one.

Reply:
I already explained in https://github.com/PowerShell/PowerShell/pull/19151#discussion_r1116029089 why it's arguably incorrect. I'm not trying to persuade you, just to clarify my opinion.

Reply:
I'm trying to figure out what to demand from other contributors so we don't waste time later. So you'd better convince me. 😃 
Right now I don't see that a ArgumentException message would be better than an ArgumentOutOfRangeException. (If I understand correctly you have no objection to the exception class type used but only to content of message.)
While I'm on vacation and have no opportunity to experiment. I'll try it a week later.

Reply:
A custom message will be absolutely better than `'paramName' must be a non-zero value. (Parameter 'paramName')`, which is what `ArgumentOutOfRangeException.ThrowIfZero` gives you.

In this particular case, the message should be something like `'paramName' must be a non-empty collection. (Parameter 'paramName')`

> you have no objection to the exception class type

I'm fine changing the type of exception to be thrown. That's a bucket 3 breaking change.

Reply:
> I took a look at the history of this API. It was created by Stephan Toub and updated by him. It is hard to suspect him of negligence. I think this is intentional since the exact API was rejected as I mentioned above. So after a while this pattern may become a frequently used one.

To be clear, `ClientWebSocketOptions.SetBuffer(Int32, Int32, ArraySegment<Byte>) ` is a public API introduced in .NET Framework 4.5. In any case, please could we should start discussion in a new issue.

Reply:
> > I took a look at the history of this API. It was created by Stephan Toub and updated by him. It is hard to suspect him of negligence. I think this is intentional since the exact API was rejected as I mentioned above. So after a while this pattern may become a frequently used one.
> 
> To be clear, `ClientWebSocketOptions.SetBuffer(Int32, Int32, ArraySegment<Byte>) ` is a public API introduced in .NET Framework 4.5. In any case, please could we continue discussion in a new issue.